src/client-side-encryption/state_machine.ts

```diff
@@ -153,176 +151,125 @@ export class StateMachine {
-            // If we are not bypassing spawning, then we should retry once on a MongoTimeoutError (server selection error)
-            if (
-              err instanceof MongoNetworkTimeoutError &&
-              mongocryptdManager &&
-              !mongocryptdManager.bypassSpawn
-            ) {
-              mongocryptdManager.spawn(() => {
-                // TODO: should we be shadowing the variables here?
-                this.markCommand(mongocryptdClient, context.ns, command, (err, markedCommand) => {
-                  if (err || !markedCommand) return callback(err);
-
-                  context.addMongoOperationResponse(markedCommand);
-                  context.finishMongoOperation();
-
-                  this.execute(executor, context, callback);
-                });
-              });
-              return;
-            }
-            return callback(err);
+            );
           }
-          context.addMongoOperationResponse(markedCommand);
-          context.finishMongoOperation();
 
-          this.execute(executor, context, callback);
-        });
+          // When we are using the shared library, we don't have a mongocryptd manager.
+          const markedCommand: Uint8Array = mongocryptdManager
+            ? await mongocryptdManager.withRespawn(
+                this.markCommand.bind(this, mongocryptdClient, context.ns, command)
+              )
+            : await this.markCommand(mongocryptdClient, context.ns, command);
 
-        return;
-      }
+          context.addMongoOperationResponse(markedCommand);
+          context.finishMongoOperation();
+          break;
+        }
 
-      case MONGOCRYPT_CTX_NEED_MONGO_KEYS: {
-        const filter = context.nextMongoOperation();
-        this.fetchKeys(keyVaultClient, keyVaultNamespace, filter, (err, keys) => {
-          if (err || !keys) return callback(err);
-          keys.forEach(key => {
+        case MONGOCRYPT_CTX_NEED_MONGO_KEYS: {
+          const filter = context.nextMongoOperation();
+          const keys = await this.fetchKeys(keyVaultClient, keyVaultNamespace, filter);
+
+          if (keys.length === 0) {
+            // This is kind of a hack.  For `rewrapManyDataKey`, we have tests that
+            // guarantee that when there are no matching keys, `rewrapManyDataKey` returns
+            // nothing.  We also have tests for auto encryption that guarantee for `encrypt`
+            // we return an error when there are no matching keys.  This error is generated in
+            // subsequent iterations of the state machine.
+            // Some apis (`encrypt`) throw if there are no filter matches and others (`rewrapManyDataKey`)
+            // do not.  We set the result manually here, and let the state machine continue.  `libmongocrypt`
+            // will inform us if we need to error by setting the state to `MONGOCRYPT_CTX_ERROR` but
+            // otherwise we'll return `{ v: [] }`.
+            result = { v: [] } as any as T;
```

Comment:
I'm ok with this given it makes `rewrapManyDataKey`'s implementation a bit cleaner.